include/envoy/network/connection.h

```diff
@@ -216,6 +220,13 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
    * @return boolean telling if the connection is currently above the high watermark.
    */
   virtual bool aboveHighWatermark() const PURE;
+
+  /**
+   * Get the socket options to be set on all new upstream ClientConnections created to forward
+   * traffic received on this connection.
+   */
+  virtual const ConnectionSocket::OptionsSharedPtr&
+  socketOptionsForUpstreamConnections() const PURE;
```

Comment:
From an interface perspective this bothers me a bit. There really shouldn't any logical association between a connection and an "upstream connection." Is there any other way we can express this? I'm wondering if it's as simple as actually just calling this `socketOptions()` to express the options for this connection, which would be used for upstream connections or any other purpose by other parts of the system. Thoughts?

Reply:
On a second thought I think this is possible if we make the distinction of the calling site/purpose on the visitor name, i.e., we inherit the socketOptions from server to the associated client connections and the client connections then call the visitor as "setClientSocketOptions()" etc. This better?

Reply:
Marginally (I will take it over what we have now), but not a ton. In the case of marking, aren't you basically trying to transfer options? Why can't we just expose options from a connection and then in other code use those options to match a new connection to an old one?

Reply:
Let me see if I understand this right.. These calls should be like (naming is elaborate to make the point):

setSocketOptionsForNewConnections() - set the options to be exposed
socketOptionsForNewConnections() - get the exposed options, if any.

and then the callback/visitor to apply the options to the new connection:

if (options) {
  options->setOptions(ConnectionSocket&);
}

I.e., there is "by definition" no confusion about the options being applied to the connection they are set at, because they are only ever to be transferred to new connections?

So this is just about naming the members and how to document them? I'm fine naming these members as you proposed in your first comment above, I'll try to make it clear in the comments that these options do not apply to the connection on which these calls are made, but only to any new connections they are transferred to.

Reply:
My main objection is that a connection is a connection. Tying anything to a connection about new connections is pretty fragile and IMO a poor abstraction. I will freely admit that this kind of conversation makes me vomit a little on the inside, but this is a bit of a stretch even for me. 😉 

So as I see it there are 2 options:
1) Have a connection expose its options, with a visitor which allows new connections to duplicate those options. It should not having anything to do with upstream, etc. Does this work for you?
2) We need to figure out some way to express options that should be used "upstream" in some other way, which means probably going outside of this interface. 

(1) is easier if that works (and I think it might?). (2) will require larger interface changes elsewhere.

Reply:
Coming in late, +1 to this.  We'll probably eventually have option we want to apply to the downstream socket as well, so keeping the API as neutrally named as possible while making sure the right options go to the right places would be ideal.